README.txt

WHAT THIS MODULE IS GOOD FOR
----------------------------
This module is provide simple general purpose AES encryption API to use in other modules.

In Drupal 6.x and 7.x it was also used for making your users passwords readable by admins. Since Drupal 8.X this
functionality was removed from the module to be implemented as a separate component/module.

REQUIREMENTS
----------------------
This module requires an implementation of AES encryption to work. Since 1.4 there are two supported implementations:
1. PHP's Mcrypt extension.
2. PHP Secure Communications Library (phpseclib)

You need to have at least one of these installed (both is fine as well).
Mcrypt is a lot faster than phpseclib (about 60 times faster according to my unscientific testing), so you might want
to use Mcrypt if you have it. But if you don't, then phpseclib is a great alternative, and the speed difference probably
won't matter in most cases.

If you don't have any of them, then read the next section below.

Also note that although this module SHOULD work on Windows, it has only been tested on Linux.

HOW TO GET AN AES IMPLEMENTATION
----------------------
If you don't have an AES implementation (you'll notice this when you install this module) then the easiest
implementation for you to get is probably the PHP Secure Communications Library (phpseclib).

To work with PhpSecLib you need to use it with Library module.

That's it! Try installing/enabling the module again.

If you want to use the Mcrypt implementation instead then you can find information on how to install it
here: http://php.net/mcrypt . Note that you most likely need to be running your own webserver in order to install
Mcrypt. If you're on a shared host you'll probably have to ask your hosting provider to install Mcrypt for you (or use
phpseclib instead).

ALTERNATIVE ENCRYPTION
----------------------
You may use your own encryption with this module. To do so, define a Plugin based on examples provided. Detailed
information is in the API.txt file .


CHANGELOG.txt
8.x-2.2 -> 8.x-2.1

  * All functions moved to class container Drupal\aes\AES.

  * Custom encryption implementation - classname should be passed as a parameter.

8.x-2.1

  * Release with procedural-style aes_encrypt/aes_decrypt function.

7.x-1.8 -> 8.x-2.x

  * Initiated porting to Drupal 8.

7.x-1.7 -> 7.x-1.8

  * Phpseclib: Integrated AES with the libraries API. PHPSeclib now preferable to be placed into libraries.

  * Fixed major bug with changing location of key file, thanks to digitalRoots. Issue: https://drupal.org/node/2188325 .

  * Minor: storing boolean data as a real boolean instead of string, by jurgenhaas.

7.x-1.6 -> 7.x-1.7

  * Fixed critical issue preventing module to normal usage by Aegir hosting system.

7.x-1.4 -> 7.x-1.6

 * By default functionality to tracking and showing user passwords is turned off.

 * Various fixes.

6.x-1.4 -> 7.x-1.4

 * Ported the module to Drupal 7.

6.x-1.3 -> 6.x-1.4

 * This module now supports the AES implementation provided by PHP Secure Communications Library (phpseclib). Which means that this module no longer requires PHP's Mcrypt extension (although it still supports it alongside phpseclib). Thanks to queen.zeal for bringing this library to my attention. Issue: http://drupal.org/node/579858 and of course many thanks to Jim Wigginton for writing this great library, which can be found here http://phpseclib.sourceforge.net/.

 * There's now an automatic trim() on decrypting data. This change is partly because the phpseclib implementation does this, so it's now also done in the mcrypt implementaton to maintain consistency between implementations, and partly because it has caused to some confusion/concern among several users (though it was not actually a bug).

 * Fixed a faulty implemented form. Thanks to rjmackay. Issue: http://drupal.org/node/507892

 * aes_encrypt and aes_decrypt will now return false if passed empty strings instead of raising errors. (Several issues has been discussing this and I believe there has been various reasons to it happening depending on your site settings and installed modules).

 * Requirements are now checked at install, thanks AlexisWilke. Issue: http://drupal.org/node/436368

 * Plus other minor changes which shouldn't affect the function/behavior of this module in any significant way.

5.x-1.2 -> 6.x-1.3
 
 * Ported the module to Drupal 6.
 
 * AES no longer alerts the user about a missing key if the user does not have the administer aes permission. This alert is now also logged into logging system.
 
 * Fixed a bug which caused AES encrypted passwords to be created even when this feature was turned off if an existing account was given a new password.

5.x-1.1 -> 5.x-1.2
 
 * Fixed a faulty use of preg_match which caused a warning in PHP 5.2.x.
 
 * A users encrypted password is now deleted along with the user.
 
 * Removed a redundant write to file when saving a key using the file storage method.
 
 * Fixed the anonymous account receiving an encrypted password in some situations.
 
 * Removed an unused menu item.

5.x-1.0 -> 5.x-1.1
 
 * Added option to use 192 or 256 bit versions of the Rijndael algorithms.
 
 * Drupals database abstraction layer is now used properly.
 
 * Added (untested) support for Windows.
 
 * AES-encrypted passwords are now stored separately in their own table instead of replacing the MD5 passwords in the users table.
 
 * Added option to show the decrypted password on another page.
 
 * Plus some other minor changes.


src/Form/AesAdminForm.php
<?php
/**
 * @file
 * Contains \Drupal\aes\Form\AesAdminForm.
 */

namespace Drupal\aes\Form;

use Drupal\Core\Form\ConfigFormBase;
use Drupal\Core\Form\FormStateInterface;
use Drupal\aes\AES;

/**
 * Provides a fields form controller.
 */
class AesAdminForm extends ConfigFormBase {

  /**
   * {@inheritdoc}
   */
  public function getFormId() {
    return 'aes_admin';
  }

  /**
   * {@inheritdoc}
   */
  public function buildForm(array $form, FormStateInterface $form_state) {
    // $config = $this->config('aes.settings')->getRawData();
    $fileStorage = \Drupal\Core\Config\FileStorageFactory::getActive();
    $config = $fileStorage->read('aes.settings');

    $phpsec_loaded = AES::load_phpsec(FALSE);

    $form = array();

    $form['aes'] = array(
      '#type' => 'fieldset',
      '#title' => t('AES settings'),
      '#collapsible' => FALSE,
    );

    $encryption_implementations = array();
    if ($phpsec_loaded) {
      $encryption_implementations['phpseclib'] = t('PHP Secure Communications Library (phpseclib)');
    }
    if (extension_loaded('mcrypt')) {
      $encryption_implementations['mcrypt'] = t('Mcrypt extension');
    }

    if (!empty($encryption_implementations['mcrypt']) && !empty($encryption_implementations['phpseclib'])) {
      $implementations_description = t('The Mcrypt implementation is faster than phpseclib and also lets you define the cipher to be used, other than that, the two core implementations are equivalent. Additional implementations might be added via plugin system.');
    }
    elseif (!empty($encryption_implementations['mcrypt']) && empty($encryption_implementations['phpseclib'])) {
      $implementations_description = t('The Mcrypt extension is the only installed implementation.') . $phpseclib_error_msg;
    }
    elseif (empty($encryption_implementations['mcrypt']) && !empty($encryption_implementations['phpseclib'])) {
      $implementations_description = t('PHP Secure Communications Library is the only installed implementation.');
    }

    /* @var \Drupal\aes\Plugin\AESPluginManager $plugin_manager */
    $plugin_manager = \Drupal::service('plugin.manager.aes');
    $plugin_definitions = $plugin_manager->getDefinitions();
    foreach ($plugin_definitions as $plugin_definition) {
      $encryption_implementations[$plugin_definition['id']] = 'Plugin: ' . $plugin_definition['label'];
    }

    if (empty($encryption_implementations)) {
      drupal_set_message(t('You do not have an AES implementation installed! For correct AES work you need an encryption library (PhpSecLib, MCrypt) or own plugin implementation. Consult READMDE.txt for more details.'), 'error');
      return array();
    }

    $form['aes']['implementation'] = array(
      '#type' => 'select',
      '#title' => t('AES implementation'),
      '#options' => $encryption_implementations,
      '#default_value' => $config['implementation'],
      '#description' => $implementations_description,
    );

    $form['aes']['cipher'] = array(
      '#type' => 'select',
      '#title' => t('Cipher'),
      '#options' => array(
        'rijndael-128' => 'Rijndael 128',
        'rijndael-192' => 'Rijndael 192',
        'rijndael-256' => 'Rijndael 256',
      ),
      '#default_value' => $config['cipher'],
      '#states' => array(
        'disabled' => array(
          ':input[name="implementation"]' => array('value' => 'phpseclib'),
        ),
      ),
    );
    $form['aes']['cipher_comment'] = array(
      '#type' => 'item',
      '#description' => t('Cipher can be chosen for MCrypt implementation only. The phpseclib implementation is locked to Rijndael 128.'),
      '#states' => array(
        'visible' => array(
          ':input[name="implementation"]' => array('value' => 'phpseclib'),
        ),
      ),
    );

    $form['aes']['key'] = array(
      '#type' => 'textfield',
      '#title' => t('Key'),
      '#description' => t("The key for your encryption system. You normally don't need to worry about this since this module will generate a key for you if none is specified. However you have the option of using your own custom key here."),
      '#required' => TRUE,
      '#default_value' => $config['key'],
    );

    $form['aes']['key_confirm'] = array(
      '#type' => 'textfield',
      '#title' => t('Confirm key'),
      '#required' => TRUE,
      '#default_value' => $config['key'],
    );

    $form['aes']['submit'] = array(
      '#type' => 'submit',
      '#value' => t('Save'),
    );

    return $form;
  }

  /**
   * {@inheritdoc}
   */
  public function validateForm(array &$form, FormStateInterface $form_state) {
    if ($form_state->getValue('key') != $form_state->getValue('key_confirm')) {
      $form_state->setErrorByName('key', t("The encryption keys didn't match."));
    }
  }

  /**
   * {@inheritdoc}
   */
  public function submitForm(array &$form, FormStateInterface $form_state) {
    \Drupal::logger('aes')->notice('Saving config...');
    // $config = $this->config('aes.settings')->getRawData();
    $fileStorage = \Drupal\Core\Config\FileStorageFactory::getActive();
    $config = $fileStorage->read('aes.settings');

    // If the cipher has changed...
    $old_cipher = $config['cipher'];
    $new_cipher = $form_state->getValue('cipher');
    if ($form_state->getValue('cipher') != $config['cipher']) {
      $config['cipher'] = $form_state->getValue('cipher');
      \Drupal\Core\Config\FileStorageFactory::getActive()->write('aes.settings', $config);

      // Get the old iv.
      $old_iv = $config['mcrypt_iv'];
      // create a new iv to match the new cipher
      AES::make_iv();
      // get the new iv
      $config = $fileStorage->read('aes.settings');
      $new_iv = $config['mcrypt_iv'];
    }

    // If the key has changed...
    if ($form_state->getValue('key') != $config['key']) {
      $config['key'] = $form_state->getValue('key');
      \Drupal\Core\Config\FileStorageFactory::getActive()->write('aes.settings', $config);

      drupal_set_message(t('Key changed.'));
      // @todo: invoke hook?
    }

    // If the implementation has changed...
    if ($form_state->getValue('implementation') != $config['implementation']) {

      $config['implementation'] = $form_state->getValue('implementation');
      \Drupal\Core\Config\FileStorageFactory::getActive()->write('aes.settings', $config);

      if ($form_state->getValue('implementation') == 'phpseclib') {
        // If we have switched to phpseclib implementation, set the cipher to 128, since it's the only one phpseclib supports.
        $config['cipher'] = 'rijndael-128';
        \Drupal\Core\Config\FileStorageFactory::getActive()->write('aes.settings', $config);

        // Create a new IV, this IV won't actually be used by phpseclib, but it's needed if the implementation is switched back to mcrypt.
        AES::make_iv(TRUE);
      }
    }
  }
}
